### ISSUE TITLE
Loss not decreasing in custom Network model - adapted from 3d_segmentation/unet.py

### ISSUE THREAD

**Issue by mgarbade** [facebookresearch/SparseConvNet#81]
I'm trying to use the SparseConvNet Code for a 3D ConvNet that has a voxelized input and output of size [BatchSize x NumChannels x Height x Width x Depth]. However the loss is not decreasing. The model looks like this:

```
class sparse_model(nn.Module):
    def __init__(self, dimension, spatialSize, reps=1, m=16, num_classes=20):
        nn.Module.__init__(self)
        self.sparseModel = scn.Sequential().add(
            scn.InputLayer(dimension, spatialSize, mode=3)).add(
            scn.SubmanifoldConvolution(dimension, 1, m, 3, False)).add(
            scn.UNet(dimension, reps, [m, 2 * m, 3 * m, 4 * m], residual_blocks=False, downsample=[2, 2])).add(
            scn.BatchNormReLU(m)).add(
            scn.SubmanifoldConvolution(3, m, num_classes, 3, False)).add(
            scn.SparseToDense(3, num_classes))

    def forward(self,x):
        x=self.sparseModel(x)
        return x
```

The only modification to the model in `examples/3d_segmentation/unet.py` is that I threw out the linear classification layer to make the network fully convolutional and replaced the `scn.OutputLayer(dimension))` with `scn.SparseToDense(3, num_classes))` since the latter gives me the right output dimensions in order to use my custom loss function ( which I successfully used before on another non-sparse network -> so that should not be the problem).

Has anyone an idea what could be going wrong?

**Comment by btgraham** [comment]
Replacing the final layer is unlikely to make much difference (better or worse).
What has happened to the scn.OutputLayer? How are you processing the network output>

**Comment by mgarbade** [comment]
Apparantly, there was something wrong with the way I was feeding the tensors to the network. I was following using [this function](https://discuss.pytorch.org/t/how-to-convert-a-dense-matrix-to-a-sparse-one/7809/3?u=mcpytorch) to convert my dense input tensor to a sparse input tensor:
```
def to_sparse(x):
    """ converts dense tensor x to sparse format """
    x_typename = torch.typename(x).split('.')[-1]
    sparse_tensortype = getattr(torch.sparse, x_typename)
    indices = torch.nonzero(x)
    if len(indices.shape) == 0:  # if all elements are zeros
        return sparse_tensortype(*x.shape)
    indices = indices.t()
    values = x[tuple(indices[i] for i in xrange(indices.shape[0]))]
    return sparse_tensortype(indices, values, x.size())
```

However this outputs a tensor of type `torch.sparse_coo`:
while SparseConvNet expects the type `SparseConvNetTensor`. I fixed that by converting the former into the latter. This is how I currently do it:
First I changed the network by kicking out the `scn.InputLayer`:
```
net = scn.Sequential().add(
    scn.SubmanifoldConvolution(dimension=3, nIn=1, nOut=16, filter_size=3, bias=False)).add(
    scn.UNet(dimension=3, reps=1, nPlanes=[16, 32], residual_blocks=False, downsample=[2, 2])).add(
    scn.SubmanifoldConvolution(dimension=3, nIn=16, nOut=20, filter_size=3, bias=False)).add(
    scn.SparseToDense(dimension=3, nPlanes=20))
```
I assume dimension=3 is correct, because the input dimension is 5 = 3 spatial dimensions + channel_dim + batch_dim? (not 100% sure here). Since I have input channel_dim=1 and batch_size=1, I can squeez out the singleton dimensions like this:
```
        # data_input is of shape [1, 1, 256, 256, 32] corresponding to [BatchSize x NumChannels x Height x Width x Depth ]

        data_input = torch.squeeze(data_input)  # -> new shape = [256, 256, 32]
```
and then use
```
        data_input_sparse = to_sparse(data_input)
        data_input_sparse = data_input_sparse.coalesce()  # Without this pytorch throws an error
        inputSpatialSize = net.input_spatial_size(torch.LongTensor([256, 256, 32]))  
        input = scn.InputBatch(3, inputSpatialSize)
        input.add_sample()
        locations = data_input_sparse.indices().t()
        vectors = data_input_sparse.values()[:, None]
        input.set_locations(locations, vectors, 0)
        predictions = net.forward(input)
        loss = cross_entropy3d(data_pred, data_label) # custom loss function
```

However this is obviously only working for batch_size = 1 and channel_size  = 1.
Whenever I try to increase the batch_size (and kick out the `torch_squeeze(...)`, the program crashes because of dimension mismatch.

I tried for example changing 
```        
inputSpatialSize = net.input_spatial_size(torch.LongTensor([256, 256, 32]))
input = scn.InputBatch(3, inputSpatialSize)
```        
to
```        
inputSpatialSize = net.input_spatial_size(torch.LongTensor([1, 1, 256, 256, 32]))
input = scn.InputBatch(5, inputSpatialSize)
```
without success.

Could you provide a toy example for inputting 3D data of arbitrary input channel_size and batch_size?
What is the difference between `spatial_size` and `dimension` in `net.input_spatial_size(...) and `scn.InputBatch(...)`?

**Comment by btgraham** [comment]
I would recommend using InputLayer instead of InputBatch. See
https://github.com/facebookresearch/SparseConvNet/blob/master/sparseconvnet/ioLayers.py#L15
